File: easy/0020_valid_parentheses.rb
class Solution

  def is_valid(s)
    stack = []

    mappings ={
      ')' => '(',
      ']' => '[',
      '}' => '{'
    }

    s.split('').each do |ch|
      if mappings[ch]
        if stack.size == 0
          return false
        end

        top = stack.pop
        if top != mappings[ch]
          return false
        end
      else
        stack << ch
      end
    end

    return true
  end

end

p Solution.new().is_valid("()")
p Solution.new().is_valid("()[]{}")
p Solution.new().is_valid("([)]")
p Solution.new().is_valid("{[]}")

File: easy/0020_valid_parentheses.py
class Solution:
    def isValid(self, s):
        stack = []

        mappings = {')' : '(', ']' : '[', '}' : '{'}

        for ch in s:
            if ch in mappings:
                if not stack:
                    return False

                top_elm = stack.pop()
                
                if top_elm != mappings[ch]:
                    return False
            else:
                stack.append(ch)

        return len(stack) == 0

print(Solution().isValid('()'))
print(Solution().isValid('()[]{}'))
print(Solution().isValid('([)]'))
print(Solution().isValid('{[]}'))